test/integration/framework/etcd.go

```diff
@@ -182,37 +182,58 @@ func EtcdMain(tests func() int) {
 	// Bail out early when -help was given as parameter.
 	flag.Parse()
 
-	before := runtime.NumGoroutine()
+	// Must be called *before* creating new goroutines.
+	goleakOpts := IgnoreBackgroundGoroutines()
+
+	goleakOpts = append(goleakOpts,
+		// lumberjack leaks a goroutine:
+		// https://github.com/natefinch/lumberjack/issues/56 This affects tests
+		// using --audit-log-path (like
+		// ./test/integration/apiserver/admissionwebhook/reinvocation_test.go).
+		// In normal production that should be harmless. We don't know here
+		// whether the test is using that, so we have to suppress reporting
+		// this leak for all tests.
+		//
+		// Both names occurred in practice.
+		goleak.IgnoreTopFunction("k8s.io/kubernetes/vendor/gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+		goleak.IgnoreTopFunction("gopkg.in/natefinch/lumberjack%2ev2.(*Logger).millRun"),
+
+		// At least one test (test/integration/quota) starts an event
+		// recorder and that gets stuck in a loop
+		// (https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L215-L231)
+		// which in the worst case loops twelve times with a delay of
+		// 10 seconds for each event before giving up.
+		//
+		// There's no good way to ignore this particular loop.
+		// Matching against Sleep is only indirect and not very
+		// precise.
+		goleak.IgnoreTopFunction("time.Sleep"),
```

Comment:
I'm not happy with this particular Ignore. However, adding some kind of forced shutdown to https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L215-L231 also doesn't seem very appealing.

I'm open for suggestions...

Reply:
add a context to 
```
 recordToSink(ctx context.Context, sink EventSink, event *v1.Event, eventCorrelator *EventCorrelator, sleepDuration time.Duration)
```

and cancel the context on Shutdown

https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L198-L200

Reply:
I am not sure whether that is possible. There are comments elsewhere (don't remember exactly where) that say that pending events will still get delivered after a shutdown, so the current behavior might be intentional. Changing it might break something?

Reply:
Besides, adding a context to `recordToSink` then implies changing the `StartEventWatcher` API - let's not.

The problem is that the sink blocks when it should simply give up:
https://github.com/kubernetes/kubernetes/blob/49ddc510bc6b57fa430f779372b71144d328995c/pkg/controller/replicaset/replica_set.go#L195

There's a TODO from 2016 about `EventSinkImpl`:

https://github.com/kubernetes/kubernetes/blame/49ddc510bc6b57fa430f779372b71144d328995c/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go#L154

Reply:
> Besides, adding a context to `recordToSink` then implies changing the `StartEventWatcher` API - let's not.

why, the context is internal and closed by `Shutdown()`, I didn't mean to expose it.

> that say that pending events will still get delivered after a shutdown

Events are ephemeral, but we can give some grace period and try to drain, WDYT?

Reply:
`eventBroadcasterImpl` doesn't have a context that it could pass to `recordToSink`. I guess it would be possible to create one just for this purpose. Let me try this and then we can ask others for their opinion whether it is a breaking change.

Reply:
Pushed an update. What I meant with "change Watch API" is that the result of `StartRecordingToSink` (a `watch.Interface`) cannot be used to shutdown the watch and the delivery callback. That is harder to change.

Reply:
This seems to have worked. I've turned the changed into a separate commit here: https://github.com/kubernetes/kubernetes/pull/115514